admin-dev/themes/new-theme/js/components/modal/progress-modal.ts

```diff
@@ -0,0 +1,499 @@
+    if (!this.processStopped) {
+      this.progressModal.progressIconContainer.innerHTML = this.getErrorIcon().outerHTML;
+    }
+  }
+
+  /** Return 401 on error */
+  private initListeners(params: ProgressModalParams)
+  {
+    this.errorModal.downloadErrorsButton.addEventListener('click', () => {
+      let csvContent = 'data:text/csv;charset=utf-8,';
+      this.errors.forEach(function(error) {
+        csvContent += error + "\r\n";
+      });
+      let link = document.createElement('a');
+      link.href = encodeURI(csvContent);
+      link.download = 'Errors.csv';
+      link.click();
+    });
+
+    this.errorModal.switchButton.addEventListener('click', () => {
+      this.modal.dialog.removeChild(this.errorModal.content);
+      this.modal.dialog.appendChild(this.progressModal.content);
+    });
+
+    this.progressModal.switchToErrorButton.addEventListener('click', () => {
+      this.modal.dialog.removeChild(this.progressModal.content);
+      this.modal.dialog.appendChild(this.errorModal.content);
+    });
+
+    this.progressModal.stopProcessingButton.addEventListener('click', () => {
+      if (params.cancelCallback) {
+        params.cancelCallback();
+        this.stopProcess();
+      }
+    });
+  }
+
+  public finishProcess()
+  {
+    if (this.processStopped) {
+      return
+    }
+
+    this.replaceStopProcessButton();
+
+    if (!this.errors.length) {
+      this.progressModal.progressIconContainer.innerHTML = this.getCompleteIcon().outerHTML;
+    }
+    this.processStopped = true;
+  }
+
+  public stopProcess()
+  {
+    if (this.processStopped) {
+      return
+    }
+
+    this.replaceStopProcessButton();
+    this.progressModal.progressIconContainer.innerHTML = this.getStopIcon().outerHTML;
+    this.processStopped = true;
+  }
+
+  private replaceStopProcessButton()
+  {
+    this.progressModal.stopProcessingButton.remove();
+    let closeButton = document.createElement('button');
+    closeButton.setAttribute('type', 'button');
+    closeButton.classList.add('btn', 'btn-primary', 'btn-lg', 'stop-processing');
+    closeButton.innerHTML = 'Close';
+    closeButton.dataset.dismiss = 'modal';
+    this.progressModal.footer.append(closeButton);
+  }
+
+
+  private getErrorIcon()
+  {
+    let errorIcon = document.createElement('span');
+    errorIcon.classList.add(
+      'material-icons',
+      'progress-warning-icon'
+    );
+    errorIcon.innerHTML = 'warning';
+    return errorIcon;
+  }
+
+  private getCompleteIcon()
+  {
+    let errorIcon = document.createElement('span');
+    errorIcon.classList.add(
+      'material-icons',
+      'progress-complete-icon'
+    );
+    errorIcon.innerHTML = 'check';
+    return errorIcon;
+  }
+
+  private getStopIcon()
+  {
+    let errorIcon = document.createElement('span');
+    errorIcon.classList.add(
+      'material-icons',
+      'progress-stop-icon'
+    );
+    errorIcon.innerHTML = 'close';
+    return errorIcon;
+  }
```

Comment:
Same suggestion for these ones which are very similar:

```suggestion
  private getWarningIcon(): HTMLElement
  {
    return this.getProgressIcon('warning');
  }

  private getCompleteIcon(): HTMLElement
  {
    return this.getProgressIcon('complete');
  }

  private getStopIcon(): HTMLElement
  {
    return this.getProgressIcon('stop');
  }

  private getProgressIcon(progressStatus: string): HTMLElement
  {
    let iconContent: string;
    switch (progressStatus) {
        case 'complete':
          iconContent = 'check';
          break;
        case 'stop':
          iconContent = 'close';
          break;
        default:
          iconContent = progressStatus;
          break;
    }

    const progressIcon = document.createElement('span');
    progressIcon.classList.add(
      'material-icons',
       `progress-${progressStatus}-icon`
    );
    progressIcon.innerHTML = iconContent;

    return errorIcon;
  }
```

Reply:
I renamed `getErrorIcon` into `getWarningIcon` on purpose, not only to simplify the implementation of `getProgressIcon` But also to emphasize the fact that we use a Warning icon on purpose even when there are errors.